### PR TITLE
[Process Exit Bounty] Implement Bounty for Processing Exits consolidated

### PR DESCRIPTION
**Pull request by souradeep-das**
This includes - 

 * Add process exit bounty for SE (https://github.com/omgnetwork/plasma-contracts/issues/661)

 * Add process exit bounty for IFE (https://github.com/omgnetwork/plasma-contracts/issues/680)

 * Static and Updatable Process Exit Bounty (https://github.com/omgnetwork/plasma-contracts/issues/685)
 
 * Anti-front running (https://github.com/omgnetwork/plasma-contracts/issues/684)

### CODE REVIEW

File: plasma_framework/contracts/src/exits/payment/routers/PaymentStandardExitRouter.sol [hunk 1]
```diff
@@ -27,6 +28,7 @@ contract PaymentStandardExitRouter is
     using PaymentChallengeStandardExit for PaymentChallengeStandardExit.Controller;
     using PaymentProcessStandardExit for PaymentProcessStandardExit.Controller;
     using BondSize for BondSize.Params;
+    using ExitBounty for ExitBounty.Params;
 
     // Initial bond size = 70000 (gas cost of challenge) * 20 gwei (current fast gas price) * 10 (safety margin)
```
Comment (thec00n) [inline plasma_framework/contracts/src/exits/payment/routers/PaymentStandardExitRouter.sol:33]: Should we adjust the value similar to INITIAL_EXIT_BOUNTY_SIZE?

File: plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouter.sol [hunk 2]
```diff
@@ -35,6 +36,7 @@ contract PaymentInFlightExitRouter is
     using PaymentDeleteInFlightExit for PaymentDeleteInFlightExit.Controller;
     using PaymentProcessInFlightExit for PaymentProcessInFlightExit.Controller;
     using BondSize for BondSize.Params;
+    using ExitBounty for ExitBounty.Params;
 
     // Initial IFE bond size = 185000 (gas cost of challenge) * 20 gwei (current fast gas price) * 10 (safety margin)
     uint128 public constant INITIAL_IFE_BOND_SIZE = 37000000000000000 wei;
```
Comment (thec00n) [inline plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouter.sol:42]: Should we update the INITIAL_IFE_EXIT_BOUNTY_SIZE?

File: plasma_framework/contracts/src/exits/payment/PaymentExitGame.sol [hunk 3]
```diff
@@ -41,12 +41,13 @@ contract PaymentExitGame is IExitProcessor, OnlyFromAddress, PaymentStandardExit
      * @notice Callback processes exit function for the PlasmaFramework to call
      * @param exitId The exit ID
      * @param token Token (ERC20 address or address(0) for ETH) of the exiting output
+     * @param processor The processExit initiator
      */
-    function processExit(uint168 exitId, uint256, address token) external onlyFrom(address(paymentExitGameArgs.framework)) {
+    function processExit(uint168 exitId, uint256, address token, address payable processor) external onlyFrom(address(paymentExitGameArgs.framework)) {
```
Comment (thec00n) [inline plasma_framework/contracts/src/exits/payment/PaymentExitGame.sol:46]: Would it be easier to read to rename this variable to `processExitInitiator`?

**Comment by thec00n** [comment]
As part of this PR could we document the mechanics around bounties and bonds somewhere and describe the rationale around them?

**Comment by boolafish** [comment]
> As part of this PR could we document the mechanics around bounties and bonds somewhere and describe the rationale around them? So who pays or receives funds during a specific stage and why.

I think it should be separated PR considering the size of PR but should be part of the effort of delivering the bounty I would like to ask for documenting this out too. Unfortunately we did not have any documentation on our bond upgrade mechanism too. If it would not be too much of a burden, it would be nice to have the doc describe both bond and bounty upgrade mechanism at once due to their similarity.

**Comment by thec00n** [comment]
I honestly have a hard time to reason about all the different scenarios where funds are transferred and if they are correct. I realize that we stall @souradeep-das pr but I would appreciate that we take more time and have more confidence that all scenarios are covered correctly.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessStandardExit.sol [hunk 4]
```diff
@@ -29,6 +29,11 @@ library PaymentProcessStandardExit {
         uint256 amount
     );
 
+    event BountyRewardFailed(
```
Comment (kevsul) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessStandardExit.sol:32]: I think this should be called `BountyReturnFailed` for consistency with the other events

File: plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouter.sol (hunk 2 above)
Comment (kevsul) [reply plasma_framework/contracts/src/exits/payment/routers/PaymentInFlightExitRouter.sol:42]: Not sure what you mean @thec00n ?
Do you mean update it to use 80 gwei as well?

**Review by boolafish — APPROVED** [review]
high level wise LGTM. please fix others comment though

**Comment by souradeep-das** [comment]
Updating it to 80 gwei for the bond will be a bit on the higher side since there is a high offset already. What number should we go for?

**Review by kevsul — APPROVED** [review]

**Pull request merged**